server/src/main/java/io/druid/segment/realtime/appenderator/StreamAppenderatorDriver.java

```diff
@@ -269,23 +268,21 @@ public Object persist(final Committer committer) throws InterruptedException
         .map(SegmentWithState::getSegmentIdentifier)
         .collect(Collectors.toList());
 
-    final ListenableFuture<SegmentsAndMetadata> publishFuture = Futures.transform(
+    final ListenableFuture<SegmentsAndMetadata> publishFuture = ListenableFutures.transformAsync(
         pushInBackground(wrapCommitter(committer), theSegments),
-        (AsyncFunction<SegmentsAndMetadata, SegmentsAndMetadata>) segmentsAndMetadata -> publishInBackground(
-            segmentsAndMetadata,
+        sam -> publishInBackground(
+            sam,
             publisher
         )
     );
-
-    return Futures.transform(
-        publishFuture,
-        (Function<SegmentsAndMetadata, SegmentsAndMetadata>) segmentsAndMetadata -> {
-          synchronized (segments) {
-            sequenceNames.forEach(segments::remove);
-          }
-          return segmentsAndMetadata;
-        }
-    );
+    return ListenableFutures.transformAsync(publishFuture, sam -> {
+      synchronized (segments) {
+        sequenceNames.forEach(segments::remove);
+      }
+      final SettableFuture<SegmentsAndMetadata> future = SettableFuture.create();
```

Comment:
Why change this to an async transform that generates a useless future, instead of keeping it as a normal transform?

Reply:
I'm not following, what do you mean by a "normal transform" here?

Reply:
I think I fixed this, but please check

Reply:
By "normal transform" I meant one that uses a regular `Function` rather than `AsyncFunction`. The new diff looks good to me now.